collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala

```diff
@@ -16,52 +19,55 @@ import scala.Predef.intWrapper
   * @define coll immutable array
   * @define Coll `ImmutableArray`
   */
-class ImmutableArray[+A] private[collection] (private val elements: scala.Array[Any])
+sealed abstract class ImmutableArray[+A]
   extends IndexedSeq[A]
     with IndexedSeqOps[A, ImmutableArray, ImmutableArray[A]]
     with StrictOptimizedSeqOps[A, ImmutableArray, ImmutableArray[A]] {
 
   def iterableFactory: SeqFactory[ImmutableArray] = ImmutableArray
 
+  /* The wrapped mutable `Array` */
+  def unsafeArray: Array[A @uncheckedVariance]
+  // uncheckedVariance should be safe: Array[A] for reference types A is covariant at the JVM level. Array[A] for
+  // primitive types A can only be widened to Array[Any] which erases to Object.
```

Comment:
No it is not safe:

~~~ scala
val rectangles: ImmutableArray[Rectangle] = ImmutableArray(rect1, rect2)
val shapes: ImmutableArray[Shape] = rectangles // assuming Rectangle is a Shape
val shapesArray: Array[Shape] = shapes.unsafeArray
shapesArray(0) = someCircle // assuming Circle is a Shape which is not a Rectangle
rectangles.foreach(rect => println(rect.height)) // ClassCastException! Circle can not be cast to Rectangle
~~~

Reply:
Did you test this? I would expect it to fail with an `ArrayStoreException` in `shapesArray(0) = someCircle`. But both this error and the one you demonstrate are due to the inherent unsoundness of covariant arrays.

This could be mitigated in `ImmutableArray` by making it invariant, but for an *immutable* array this should not be necessary. The JVM array actually *is* covariant and.  If you use it in a mutable way all bets are off. That's why it is called *unsafe* in the first place. Any immutable use should be sound.